// @wix/animations-kit/dist/esm/definitions/animation/mouse/tilt3DMouse.js
import {
    mapRange
} from '@wix/wow-utils';
import {
    getMouseTransitionEasings
} from '../../../utils/animationsUtils';
export const name = 'Tilt3DMouse';
export const properties = {
    schema: {},
};
export function animate(_kit, elements, _duration, _delay, {
    inverted = false,
    angle = 5,
    perspective = 800,
    currentRotation = 0,
    transitionDuration,
    transitionEasing = 'linear',
} = {}) {
    const element = elements[0];
    const invert = inverted ? -1 : 1;
    if (transitionDuration) {
        element.style.transition = `transform ${transitionDuration}ms ${getMouseTransitionEasings(transitionEasing)}`;
    }
    return {
        progress(progressX, progressY) {
            // if progressX === 0, rotateX === -angle, if progressX === 0.5, rotateX === 0, if progressX === 1, rotateX === angle
            const rotateX = mapRange(0, 1, angle, -angle, progressY) * invert;
            const rotateY = mapRange(0, 1, -angle, angle, progressX) * invert;
            const transform = `perspective(${perspective}px) rotateX(${rotateX}deg) rotateY(${rotateY}deg) rotate(${currentRotation}deg)`;
            element.style.transform = transform;
        },
        kill() {
            element.style.transform = '';
            element.style.transition = '';
        },
    };
}
